README.md
# express-humans

![stability-stable](https://img.shields.io/badge/stability-stable-green.svg)

[![npm version][version-badge]][version-url]
[![Known Vulnerabilities][vulnerabilities-badge]][vulnerabilities-url]
[![dependency status][dependency-badge]][dependency-url]
[![devdependency status][devdependency-badge]][devdependency-url]
[![build status][build-badge]][build-url]
[![Code Climate][maintainability-badge]][maintainability-url]
[![downloads][downloads-badge]][downloads-url]

[![NPM][npm-stats-badge]][npm-stats-url]

> Express middleware for generating a humans.txt or responding with an existing file.

## Usage

First run :

`npm install express-humans` (both npm and Yarn locks are supported)

then :

```javascript
const humans = require('express-humans');
const express = require('express');

const app = express();

app.use(humans({
  team: [
    {
      'Original developer': 'Hayden Bleasel',
      Twitter: '@haydenbleasel'
    }, {
      Maintainer: 'Alexis Paques',
      Github: '@AlexisTM'
    }
  ],
  thanks: [
    'Node',
    'Gulp'
  ],
  site: {
    'Standards': 'HTML5, CSS3',
    'Components': 'jQuery, Normalize.css',
    'Softwares': 'Atom, SublimeText'
  },
  note: 'Built with love by Hayden Bleasel.'
}));

app.listen(3000);
```

or :

```javascript
// Using the path of the humans.txt file
app.use(humans('./humans.txt'));
```

## API

`humans(options)`

* `options` : `String` or `Object`
  * if `String` : the path of the humans.txt file to serve
  * if `Object` : [the `humans-generator` package config](https://www.npmjs.com/package/humans-generator)

## Development

```bash
npm run build # Or "yarn run build"
npm test # Or "yarn test"
```

## Credits

Inspired of [express-robots](https://www.npmjs.com/package/express-robots).

## License

MIT License

Copyright (c) 2017-2018 **Nicolas COUTIN**

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.

[version-badge]: https://img.shields.io/npm/v/express-humans.svg
[version-url]: https://www.npmjs.com/package/express-humans
[vulnerabilities-badge]: https://snyk.io/test/npm/express-humans/badge.svg
[vulnerabilities-url]: https://snyk.io/test/npm/express-humans
[dependency-badge]: https://david-dm.org/ilshidur/express-humans.svg
[dependency-url]: https://david-dm.org/ilshidur/express-humans
[devdependency-badge]: https://david-dm.org/ilshidur/express-humans/dev-status.svg
[devdependency-url]: https://david-dm.org/ilshidur/express-humans#info=devDependencies
[build-badge]: https://travis-ci.org/Ilshidur/express-humans.svg
[build-url]: https://travis-ci.org/Ilshidur/express-humans
[maintainability-badge]: https://api.codeclimate.com/v1/badges/88f5bf63731c1cc35f9e/maintainability
[maintainability-url]: https://codeclimate.com/github/Ilshidur/express-humans/maintainability
[downloads-badge]: https://img.shields.io/npm/dt/express-humans.svg
[downloads-url]: https://www.npmjs.com/package/express-humans
[npm-stats-badge]: https://nodei.co/npm/express-humans.png?downloads=true&downloadRank=true
[npm-stats-url]: https://nodei.co/npm/express-humans


CHANGELOG.md
# league-sprites - change log

# 2.0.1 (Aug. 3rd 2018)

## **Minor changes :**

* Removed unused dependencies

# 2.0.0 (Aug. 3rd 2018)

## **Major changes :**

* Migrate to native ES8 code, without gulp & Babel

# 1.0.3 (Oct. 9th 2017)

## **Minor fixes :**

* Fix vulnerable dependencies

# 1.0.2 (Sep. 5th 2017)

## **Minor fixes :**

* Dependencies upgrade
* Typo fix in README

# 1.0.1 (Aug. 13th 2017)

## **Minor fixes :**

* Travis-CI package build
* Usage & API in README.md
* Stability shield in README.md

# 1.0.0 (July. 10th 2017)

First release.


test/index.js
const fs = require('fs');
const path = require('path');
const expect = require('chai').expect;
const supertest = require('supertest');
const figlet = require('figlet');
const humans = require('..');

const route = '/humans.txt';
const header = figlet.textSync('Humans.txt') + '\n\n';

describe('express-humans :', function() {
  it('should output empty string if no config', function(done) {
    supertest(humans())
      .get(route)
      .end(function(err, res) {
        expect(res.status).to.equal(404);
        expect(res.headers['content-type']).to.equal('text/plain; charset=utf-8');
        expect(res.text).to.equal('404 : Not Found');
        done();
      });
  });

  it('should work with an input file', function(done) {
    supertest(humans(path.join(__dirname, 'fixtures/humans.txt')))
      .get(route)
      .end(function(err, res) {
        expect(res.status).to.equal(200);
        expect(res.headers['content-type']).to.equal('text/plain; charset=utf-8');
        expect(res.text).to.equal('/* TEAM */\n\nChef: Ilshidur\n\n\n/* THANKS */\n\nName: NPM\n\n\n/* SITE */\n\nLast update: never\n');
        done();
      });
  });

  describe('should work with an input config :', function() {

    it('team - string', function(done) {
      supertest(humans({
        team: 'test'
      }))
        .get(route)
        .end(function(err, res) {
          expect(res.status).to.equal(200);
          expect(res.headers['content-type']).to.equal('text/plain; charset=utf-8');
          expect(res.text).to.equal(header + '/* TEAM */\ntest\n');
          done();
        });
    });
    it('team - array', function(done) {
      supertest(humans({
        team: [
          'test'
        ]
      }))
        .get(route)
        .end(function(err, res) {
          expect(res.status).to.equal(200);
          expect(res.headers['content-type']).to.equal('text/plain; charset=utf-8');
          expect(res.text).to.equal(header + '/* TEAM */\ntest\n');
          done();
        });
    });
    it('thanks - string', function(done) {
      supertest(humans({
        thanks: 'test'
      }))
        .get(route)
        .end(function(err, res) {
          expect(res.status).to.equal(200);
          expect(res.headers['content-type']).to.equal('text/plain; charset=utf-8');
          expect(res.text).to.equal(header + '/* THANKS */\ntest\n');
          done();
        });
    });
    it('thanks - array', function(done) {
      supertest(humans({
        thanks: [
          'test'
        ]
      }))
        .get(route)
        .end(function(err, res) {
          expect(res.status).to.equal(200);
          expect(res.headers['content-type']).to.equal('text/plain; charset=utf-8');
          expect(res.text).to.equal(header + '/* THANKS */\ntest\n');
          done();
        });
    });
    it('site - string', function(done) {
      supertest(humans({
        site: 'test'
      }))
        .get(route)
        .end(function(err, res) {
          expect(res.status).to.equal(200);
          expect(res.headers['content-type']).to.equal('text/plain; charset=utf-8');
          expect(res.text).to.equal(header + '/* SITE */\ntest\n');
          done();
        });
    });
    it('site - array', function(done) {
      supertest(humans({
        site: [
          'test'
        ]
      }))
        .get(route)
        .end(function(err, res) {
          expect(res.status).to.equal(200);
          expect(res.headers['content-type']).to.equal('text/plain; charset=utf-8');
          expect(res.text).to.equal(header + '/* SITE */\ntest\n');
          done();
        });
    });
    it('all params', function(done) {
      supertest(humans({
        team: [
          'test'
        ],
        thanks: [
          'test'
        ],
        site: [
          'test'
        ]
      }))
        .get(route)
        .end(function(err, res) {
          expect(res.status).to.equal(200);
          expect(res.headers['content-type']).to.equal('text/plain; charset=utf-8');
          expect(res.text).to.equal(header + '/* TEAM */\ntest\n\n/* THANKS */\ntest\n\n/* SITE */\ntest\n');
          done();
        });
    });

  });

  it('should throw an error if the path is incorrect', function(done) {
    supertest(humans(path.join(__dirname, 'fixtures/humans.txtINCORRECT')))
      .get(route)
      .end(function(err, res) {
        expect(res.status).to.equal(500);
        expect(res.text).to.equal('500 : No humans.txt found !');
        done();
      });
  });
});


src/index.js
const fs = require('fs');
const express = require('express');
const humans = require('humans-generator');

const route = '/humans.txt';

/**
 * This express middleware loads the file content (or generates the humans.txt
 * content, depending on its use). Once done, it will always serve the same file
 * from the memory.
 * To avoid this behavior and serve the content dynamically, pass the
 * 'dynamic: true' parameter in the middleware options.
 */
function middleware(config, opts) {
  // TODO: 'opts.dynamic'

  async function getContent() {
    let content = '';

    if (config) {
      if (typeof config === 'string') {
        // Config is file path
        content = fs.readFileSync(config, 'utf8');
      } else {
        // Config is humans.txt as an object
        content = await generateContent(config);
      }
    }

    return content;
  }

  const app = express();

  app.get(route, function(req, res, next) {
    getContent()
      .then((content) => {
        if (content === '') {
          res.header('Content-Type', 'text/plain');
          res.status(404).send('404 : Not Found');
        } else {
          res.header('Content-Type', 'text/plain');
          res.status(200).send(content);
        }
      })
    .catch((err) => {
      res.status(500).send('500 : No humans.txt found !');
    });
  });

  return app;
}

function generateContent(config) {
  return new Promise((resolve, reject) => {
    humans(config, (err, content) => {
      if (err) {
        reject(err);
      } else {
        resolve(content.join('\n'));
      }
    });
  });
}

module.exports = middleware;
